python3Packages.labgrid: init at 0.3.1

Emantor opened the pull request:
###### Motivation for this change
Labgrid is an embedded control and testing library, this package
packages the minimum to let the client and core-library work.

Labgrid also has a check for a specific pyserial port we maintain in the project.
Not sure how to deal with this and whether packaging the fork for nixos is fine.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Emantor committed:
python3Packages.labgrid: init at 0.3.1

Labgrid is an embedded control and testing library, this package
packages the minimum to let the client and core-library work.

Signed-off-by: Rouven Czerwinski <rouven@czerwinskis.de>

SuperSandro2000 commented on pkgs/development/python-modules/labgrid/default.nix:
```diff
@@ -0,0 +1,70 @@
+    sha256 = "0zyf4cpazkxlbd1acmx3i1awi4kpi43g5qdvz41w2ls66h5jwf0f";
+  };
+
+  nativeBuildInputs = [ setuptools_scm ];
```
```suggestion
  nativeBuildInputs = [ setuptools-scm ];
```

mweinelt replied:
We are trying to normalize python package names according to https://www.python.org/dev/peps/pep-0503/#normalized-names.

SuperSandro2000 commented on pkgs/development/python-modules/labgrid/default.nix:
```diff
@@ -0,0 +1,70 @@
+  checkInputs = [
+    mock
+    psutil
+    pytest
```
```suggestion
    pytestCheckHook
```

SuperSandro2000 commented on pkgs/development/python-modules/labgrid/default.nix:
```diff
@@ -0,0 +1,70 @@
+    pytest-mock
+  ];
+
+  checkPhase = ''
+    pytest -k "not docker and not pyvisa and not sshmanager and not graphviz" tests
+  '';
```
```suggestion
  disabledTests = [
    "docker"
    "graphviz"
    "pyvisa"
    "sshmanager"
  ];
```

SuperSandro2000 commented on pkgs/development/python-modules/labgrid/default.nix:
```diff
@@ -0,0 +1,70 @@
+, pexpect
+, psutil
+, pyserial
+, pytest
```
```suggestion
, pytestCheckHook
```

mweinelt commented:
> Labgrid also has a check for a specific pyserial port we maintain in the project.
> Not sure how to deal with this and whether packaging the fork for nixos is fine.

If required you can maintain a fork of pyserial in the labgrid project directory.

The `gitAndTools.gitstatus` package has a similiar relationship to a libgit2 fork if you're interested in how it's done.

https://github.com/NixOS/nixpkgs/blob/eccdb601c1662e96951c1ec8e23791c9d5ef8814/pkgs/applications/version-management/git-and-tools/gitstatus/default.nix#L14

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/version-management/git-and-tools/gitstatus/romkatv_libgit2.nix

worldofpeace commented:
We merged the patches into our pyserial. Upstream should really accept them, and if they don't then eventually downstreams are just going to roll those patches on their own.

worldofpeace commented on pkgs/development/python-modules/labgrid/default.nix:
```diff
@@ -0,0 +1,72 @@
+  };
+
+  nativeBuildInputs = [ setuptools-scm ];
+  propagatedBuildInputs = [
```
```suggestion
  
  propagatedBuildInputs = [
```

worldofpeace commented on pkgs/development/python-modules/labgrid/default.nix:
```diff
@@ -0,0 +1,72 @@
+, pyyaml
+, requests
+, setuptools-scm
+, xmodem }:
```
```suggestion
, xmodem 
}:
```

worldofpeace commented:
If you can tell from my suggestions I can't find anything wrong with this expression.

Emantor commented:
> If you can tell from my suggestions I can't find anything wrong with this expression.

I fixed up both suggestions and added a patch to remove the pyserial check (which uses version and won't work for NixOS any longer, since NixOS applies the patches downstream). Thanks for the review, this should be good to go now.

mweinelt commented:
This looks a bit odd, but it does not seem to cause issues. Do you know whats happening here? @Emantor 

```
=============== 201 passed, 31 skipped, 27 deselected in 36.30s ================
An exception occured during cleanup, call the cleanup() method on targets yourself to handle exceptions explictly.
Error: {} [Errno 2] No such file or directory: 'ssh'
An exception occured during cleanup, call the cleanup() method on targets yourself to handle exceptions explictly.
Error: {} [Errno 2] No such file or directory: 'ssh'
An exception occured during cleanup, call the cleanup() method on targets yourself to handle exceptions explictly.
Error: {} child has vanished
Finished executing pytestCheckPhase
```

Apart from that I'm also content with the state of this.

Emantor commented:
> This looks a bit odd, but it does not seem to cause issues. Do you know whats happening here? @Emantor
> 
> ```
> =============== 201 passed, 31 skipped, 27 deselected in 36.30s ================
> An exception occured during cleanup, call the cleanup() method on targets yourself to handle exceptions explictly.
> Error: {} [Errno 2] No such file or directory: 'ssh'
> An exception occured during cleanup, call the cleanup() method on targets yourself to handle exceptions explictly.
> Error: {} [Errno 2] No such file or directory: 'ssh'
> An exception occured during cleanup, call the cleanup() method on targets yourself to handle exceptions explictly.
> Error: {} child has vanished
> Finished executing pytestCheckPhase
> ```
> 
> Apart from that I'm also content with the state of this.

Yes, we do have some tests which setup a mocked SSH connection, but during teardown we try to cleanup all connections. The cleanup tries to use `ssh`, but this is not available during the test run on NixOS. The cleanup code was added after the tests were introduced and almost every other platform does have `ssh` available while running the tests. IMO this is fine to ignore for now, I'll fix the tests for the next release so this should vanish on the next update.
